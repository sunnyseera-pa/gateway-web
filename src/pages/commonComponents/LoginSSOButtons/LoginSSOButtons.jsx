/** @jsx jsx */
import { jsx } from '@emotion/react';
import React, { useEffect, useReducer, Fragment } from 'react';
import { Row, Col } from 'react-bootstrap';
import PropTypes from 'prop-types';
import { chunk } from 'lodash';
import '../CommonComponents.scss';
import * as styles from './LoginSSoButtons.styles';
import lastChoiceSVG from '../../../images/lastChoice.svg';

const reducer = (ssoBtnsState, lastChoice) => {
	return ssoBtnsState.map(value => {
		value.active = false;
		if (value.id === lastChoice) {
			value.active = true;
		}
		return value;
	});
};

function LoginSSOButtons({ ssoBtnsConfig, communityLink, clickHandler, lastChoice }) {
	const [ssoBtnsState, dispatch] = useReducer(reducer, ssoBtnsConfig);

	useEffect(() => {
		if (lastChoice) {
			dispatch(lastChoice);
		}
	}, [lastChoice]);

	return (
		<div>
			{chunk(ssoBtnsState, 2).map((arr, index) => (
				<Fragment key={index}>
					<br />
					<Row className='mt-2'>
						<Col sm={0} lg={1} />
						{arr.map((value, i) => (
							<Col sm={6} lg={5} key={value.id + i} className='mt-1'>
								<div className='gray800-14'>
									<button
										data-testid={value.id}
										className='btn btn-outline-secondary btn-block'
										css={styles.btn}
										onClick={() => clickHandler(value.id, value.authURL)}>
										<img src={value.img} css={styles.btnImg} alt={value.id} />
										&nbsp; {value.text}
										{value.active ? (
											<img src={lastChoiceSVG} css={styles.lastChoiceBtnImg} data-testid={value.id + '-lastChoice'} alt='lastChoice' />
										) : null}
									</button>
								</div>
							</Col>
						))}
						<Col sm={0} lg={1} key={'col-1-' + index} />
					</Row>
				</Fragment>
			))}
			<div css={styles.supportLink}>
				<Row className='mt-5 '>
					<Col sm={0} lg={1} />
					<Col sm={6} lg={5}>
						<span>
							{' '}
							<a
								target='_blank'
								href={`${communityLink}/t/how-to-submit-a-feature-request-or-feedback/1`}
								data-testid='communityLink'
								rel='noopener noreferrer'>
								Suggest another Indentity Provider
							</a>
						</span>
					</Col>
					<Col sm={6} lg={5}>
						{lastChoice ? (
							<span>
								<img src={lastChoiceSVG} css={styles.btnImg} data-testid='lastChoiceNote' alt='lastChoice' /> Last time you clicked this
								button
							</span>
						) : null}
					</Col>
					<Col sm={0} lg={1} />
				</Row>
			</div>
		</div>
	);
}

LoginSSOButtons.propTypes = {
	clickHandler: PropTypes.func.isRequired,
	communityLink: PropTypes.string.isRequired,
	lastChoiceSVG: PropTypes.string,
	ssoBtnsConfig: PropTypes.array.isRequired,
};

export default LoginSSOButtons;
