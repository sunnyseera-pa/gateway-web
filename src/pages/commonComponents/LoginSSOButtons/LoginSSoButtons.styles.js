import { css } from '@emotion/react';

export const btnImg = css`
    width: 20px;
    height: 20px;
    float: left;
`;
export const btn = css`
    text-align: left;
`;

export const supportLink = css`
    font-size: 12px;
`;
export const lastChoiceBtnImg = css`
    margin-right: -10px;
    float: right;
    width: 20px;
`;
