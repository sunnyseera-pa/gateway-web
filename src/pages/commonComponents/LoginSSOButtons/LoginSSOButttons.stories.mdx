<!--- Checkbox.stories.mdx -->

import { Meta, Story, Canvas } from '@storybook/addon-docs';

import LoginSSOButtons from './LoginSSOButtons';
import { ssoBtnsConfig } from '../../../configs/ssoBtnsConfig';

<Meta
	title='LoginSSOButtons'
	component={LoginSSOButtons}
	args={{
		communityLink: 'test',
		lastChoice: 'google',
		ssoBtnsConfig: ssoBtnsConfig,
	}}
/>

<Canvas>
	<Story name='Basic'>
		{props => {
			const [lastChoice, setLastChoice] = React.useState('');
			const handleClick = (id, authURL) => {
				setLastChoice(id);
			};
			return <LoginSSOButtons {...props} clickHandler={handleClick} lastChoice={lastChoice} />;
		}}
	</Story>
</Canvas>
