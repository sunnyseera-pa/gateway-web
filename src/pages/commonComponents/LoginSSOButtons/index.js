import LoginSSOButtons from './LoginSSOButtons';

export default LoginSSOButtons;
