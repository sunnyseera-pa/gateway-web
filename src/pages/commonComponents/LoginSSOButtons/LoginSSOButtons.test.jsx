import React from 'react';
import LoginSSOButtons from './index';
import _ from 'lodash';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

const props = {
    ssoBtnsConfig: [
        { id: 'google', text: 'Sign in with Google', authURL: '/auth/google', img: 'googleIcon', active: false },
        { id: 'openAthens', text: 'Sign in with institution', authURL: '/auth/oidc', img: 'instIcon', active: false },
        { id: 'azure', text: 'Sign in with Microsoft', authURL: '/auth/azure', img: 'microsoftIcon', active: false },
    ],
    communityLink: 'https://test',
};
let wrapper;
let handlerFn;

describe('Given the LoginSSOButtons component', () => {
    describe('When it is rendered', () => {
        beforeAll(() => {
            wrapper = render(<LoginSSOButtons {...props} />);
        });

        it('Then matches the previous snapshot', () => {
            expect(wrapper.container).toMatchSnapshot();
        });

        it('Then all buttons rendered', () => {
            props.ssoBtnsConfig.map(value => {
                expect(screen.getByTestId(value.id)).toBeTruthy();
            });
        });

        it('Then lastChoice Messgage should not be rendered', () => {
            expect(screen.queryByTestId('lastChoiceNote')).toBeNull();
        });

        it('Then Suggest another Indentity Provider link should  be rendered', () => {
            expect(screen.getByTestId('communityLink')).toHaveAttribute(
                'href',
                `${props.communityLink}/t/how-to-submit-a-feature-request-or-feedback/1`
            );
        });
    });

    describe('And lastchoice is google', () => {
        it('Then has the correct output button with lastChoiceSVG icon', () => {
            const { rerender, container } = wrapper;
            handlerFn = jest.fn();
            rerender(<LoginSSOButtons {...props} lastChoice='google' clickHandler={handlerFn} />);
            props.ssoBtnsConfig.map(value => {
                expect(screen.getByTestId(value.id)).toBeTruthy();
                if (value.id === 'google') {
                    expect(screen.getByTestId(value.id + '-lastChoice')).toBeTruthy();
                } else {
                    expect(screen.queryByTestId(value.id + '-lastChoice')).toBeNull();
                }
            });
        });
        it('Then has Last Time you clicked this button text', () => {
            expect(screen.getByTestId('lastChoiceNote')).toBeTruthy();
        });
        it('Then should call clickHandlder function when google button clicked', () => {
            fireEvent.click(screen.getByTestId('google'));
            expect(handlerFn.mock.calls.length).toBe(1);
            expect(handlerFn.mock.calls[0][0]).toBe('google');
            expect(handlerFn.mock.calls[0][1]).toBe('/auth/google');
        });
    });
    describe('And lastchoice is azure', () => {
        it('Then has the correct output button with lastChoiceSVG icon', () => {
            const { rerender, container } = wrapper;
            handlerFn = jest.fn();
            rerender(<LoginSSOButtons {...props} lastChoice='azure' clickHandler={handlerFn} />);
            props.ssoBtnsConfig.map(value => {
                expect(screen.getByTestId(value.id)).toBeTruthy();
                if (value.id === 'azure') {
                    expect(screen.getByTestId(value.id + '-lastChoice')).toBeTruthy();
                } else {
                    expect(screen.queryByTestId(value.id + '-lastChoice')).toBeNull();
                }
            });
        });
        it('Then has Last Time you clicked this button text', () => {
            expect(screen.getByTestId('lastChoiceNote')).toBeTruthy();
        });
        it('Then should call clickHandlder function when azure button clicked', () => {
            fireEvent.click(screen.getByTestId('azure'));
            // expect(props.clickHandler.mock.calls.length).toBe(1);
            expect(handlerFn.mock.calls[0][0]).toBe('azure');
            expect(handlerFn.mock.calls[0][1]).toBe('/auth/azure');
        });
    });
});
