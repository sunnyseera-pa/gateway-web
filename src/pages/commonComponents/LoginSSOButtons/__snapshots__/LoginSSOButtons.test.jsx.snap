// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the LoginSSOButtons component When it is rendered Then matches the previous snapshot 1`] = `
<div>
  <div>
    <br />
    <div
      class="mt-2 row"
    >
      <div
        class="col-lg-1"
      />
      <div
        class="mt-1 col-lg-5 col-sm-6"
      >
        <div
          class="gray800-14"
        >
          <button
            class="btn btn-outline-secondary btn-block css-ayanwq-LoginSSOButtons"
            data-testid="google"
          >
            <img
              alt="google"
              class="css-10nh4b5-LoginSSOButtons"
              src="googleIcon"
            />
              
            Sign in with Google
          </button>
        </div>
      </div>
      <div
        class="mt-1 col-lg-5 col-sm-6"
      >
        <div
          class="gray800-14"
        >
          <button
            class="btn btn-outline-secondary btn-block css-ayanwq-LoginSSOButtons"
            data-testid="openAthens"
          >
            <img
              alt="openAthens"
              class="css-10nh4b5-LoginSSOButtons"
              src="instIcon"
            />
              
            Sign in with institution
          </button>
        </div>
      </div>
      <div
        class="col-lg-1"
      />
    </div>
    <br />
    <div
      class="mt-2 row"
    >
      <div
        class="col-lg-1"
      />
      <div
        class="mt-1 col-lg-5 col-sm-6"
      >
        <div
          class="gray800-14"
        >
          <button
            class="btn btn-outline-secondary btn-block css-ayanwq-LoginSSOButtons"
            data-testid="azure"
          >
            <img
              alt="azure"
              class="css-10nh4b5-LoginSSOButtons"
              src="microsoftIcon"
            />
              
            Sign in with Microsoft
          </button>
        </div>
      </div>
      <div
        class="col-lg-1"
      />
    </div>
    <div
      class="css-k2m949-LoginSSOButtons"
    >
      <div
        class="mt-5  row"
      >
        <div
          class="col-lg-1"
        />
        <div
          class="col-lg-5 col-sm-6"
        >
          <span>
             
            <a
              data-testid="communityLink"
              href="https://test/t/how-to-submit-a-feature-request-or-feedback/1"
              rel="noopener noreferrer"
              target="_blank"
            >
              Suggest another Indentity Provider
            </a>
          </span>
        </div>
        <div
          class="col-lg-5 col-sm-6"
        />
        <div
          class="col-lg-1"
        />
      </div>
    </div>
  </div>
</div>
`;
